components/tools/OmeroWeb/omeroweb/webgateway/views.py

```diff
@@ -2514,3 +2571,338 @@ def object_table_query(request, objtype, objid, conn=None, **kwargs):
     tableData['parentId'] = ann['parentId']
     tableData['addedOn'] = ann['addedOn']
     return tableData
+
+
+def build_url(request, name, api_version, **kwargs):
+    kwargs['api_version'] = api_version
+    return request.build_absolute_uri(
+        reverse(name, kwargs=kwargs))
+
+
+@json_response
+def api_versions(request, **kwargs):
+    """
+    Base url of the webgateway json api.
+    """
+    versions = []
+    for v in settings.API_VERSIONS:
+        versions.append({
+            'version': v,
+            'base_url': build_url(request, 'api_base', v)
+        })
+    return {'data': versions}
+
+
+@json_response
+def api_base(request, api_version=None, **kwargs):
+    """
+    Base url of the webgateway json api for a specified version.
+    """
+    v = api_version
+    rv = {'projects_url': build_url(request, 'api_projects', v),
+          'token_url': build_url(request, 'api_token', v),
+          'servers_url': build_url(request, 'api_servers', v),
+          'login_url': build_url(request, 'api_login', v),
+          'save_url': build_url(request, 'api_save', v),
+          'schema_url': OME_SCHEMA_URL}
+    return rv
+
+
+@json_response
+def api_token(request, api_version, **kwargs):
+    """
+    Provides CSRF token for current session
+    """
+    token = csrf.get_token(request)
+    return {'data': token}
+
+
+@json_response
+def api_servers(request, api_version, **kwargs):
+    """
+    Lists the available servers to connect to
+    """
+    servers = []
+    for i, obj in enumerate(Server):
+        s = {'id': i + 1,
+             'host': obj.host,
+             'port': obj.port
+             }
+        if obj.server is not None:
+            s['server'] = obj.server
+        servers.append(s)
+    return {'data': servers}
+
+
+class LoginView(View):
+    """
+    Webgateway Login - Subclassed by WebclientLoginView
+    """
+
+    form_class = LoginForm
+    useragent = 'OMERO.webapi'
+
+    def get(self, request, api_version=None):
+        """ Simply return a message to say GET not supported """
+        return JsonResponse({"message":
```

Comment:
I guess that using ```require_POST``` would be the same as not having a ```get``` method on the ```LoginForm``` class. That would probably be fine, and it's what we're doing in a bunch of other places. It's slightly nicer to have a helpful message (particularly if exploring the API in a browser). I'll certainly change the status to 405 and we can decide later if we want any placeholders for GET. e6a942f